test/dynamo/test_higher_order_ops.py

```diff
@@ -282,13 +282,183 @@ def false_fn2(x, y):
-                "cond_false_0.cond_true_0",
+                "cond_true_1",
+                "cond_false_1",
+                "cond_false_1.cond_false_0",
+                "cond_false_1.cond_true_0",
             },
         )
 
+    def test_cond_graph_break_in_one_branch(self):
+        backend = EagerAndRecordGraphs()
+        cnt = CompileCounterWithBackend(backend)
+
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("buffer", torch.ones(6, 4))
+
+            def forward(self, x):
+                def true_fn(x):
+                    self.buffer += 1
+                    return self.buffer.sum() + x.sum()
+
+                def false_fn(x):
+                    return (x - 1).sum()
+
+                return control_flow.cond(x.shape[0] > 4, true_fn, false_fn, [x])
+
+        mod_for_compile = torch.compile(Foo(), backend=cnt, dynamic=True)
+        mod_for_eager = Foo()
+
+        self.assertEqual(
+            mod_for_compile(torch.ones(6, 4)), mod_for_eager(torch.ones(6, 4))
+        )
+
+    def test_cond_free_variable_in_both_branches(self):
+        backend = EagerAndRecordGraphs()
+        cnt = CompileCounterWithBackend(backend)
+
+        z = torch.ones(4, 4)
+
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("buffer", torch.ones(6, 4))
+
+            def forward(self, x, y):
+                def true_fn(x):
+                    return x.sum() + self.buffer.sum() + z.sum()
+
+                def false_fn(x):
+                    return x.sum() - self.buffer.sum() - z.sum()
+
+                return control_flow.cond(y, true_fn, false_fn, [x])
+
+        mod_for_compile = torch.compile(
+            Foo(), backend=cnt, dynamic=True, fullgraph=True
+        )
+        mod_for_eager = Foo()
+
+        self.assertEqual(
+            mod_for_compile(torch.tensor(True), torch.tensor(5)),
+            mod_for_eager(torch.tensor(True), torch.tensor(5)),
+        )
+
+        for node in backend.graphs[0].graph.nodes:
+            if node.op == "call_function" and node.target == control_flow.cond:
+                _, _, _, operands = node.args
+                # Each branch takes 5 inputs (x, true_buffer, true_z, false_buffer, false_z)
+                self.assertEqual(len(operands), 5)
+            if node.op == "get_attr":
+                if str(node.target) in ("cond_true_0, cond_false_0"):
+                    num_placeholders = len(
+                        [
+                            node
+                            for node in getattr(
+                                backend.graphs[0], str(node.target)
+                            ).graph.nodes
+                            if node.op == "placeholder"
+                        ]
+                    )
+                    self.assertEqual(num_placeholders, 5)
```

Comment:
I'm a bit nervous about free variables in different branches of cond, so we should add more test cases around those. What I want to see are things like:

- true_fn(x) captures (a, c, d), false_fn(x) captures (a, b, d). We want to make sure that the free variable lifting logic rewrites both true_fn and false_fn to accept (a, c, d, b) (or something reasonable). Furthermore, we should use self.assertExpectedInline so that we can actually have what the graphs look like in the code so we know what we're working with.
- true_fn(x) captures (a, c, b), false_fn(x) captures (b, a, c). The result is that Dynamo should rewrite both branches to accept (x, a, c, b) (or something reasonable). Furthermore, we should use self.assertExpectedInline so that we can actually have what the graphs look like in the code so we know what we're working with.

This is a bit of a mini-algorithms problem, given two lists of free variables, how do we unify them so that both functions accept the list of all of the variables in the same order.

Bonus points if we can write these tests in a more compact form so that we don't need to write 50 LOC to test each of these cases, because there are probably more edge cases to test than the ones I've listed above